### ISSUE TITLE
Responsive Images

### ISSUE THREAD

**Issue by IsaiahJTurner** [kean/Nuke#446]
We have a server that provides small (50px), medium (300px), and large (1000px) sizes of images. We display different sizes in different places of the app.

When we create an image view in the app, we know what image size is ideal for display. We fetch that size from the server using Nuke and it loads and caches just fine.

We'd like to further improve loading speeds by displaying the smaller size from the cache if available while loading a larger size from the server.

Further, if the large size is available in the cache and we are attempting to display a small size, we'd like to cache that size separately for performance reasons by simply downsizing the large image without making a request to the server. Think, the native Photos app. Apple caches and stores thumbnails for all your photos in separate files so they don't need to load the full size image when scrolling thousands of photos very quickly.

Is this possible with Nuke? If so, could you point me in the right direction with regards to what needs to be implemented/adjusted/customized?

**Comment by kean** [comment]
Hi,

To go from low to high resolution you can use [ImagePublisher](https://github.com/kean/ImagePublisher). I think [this](https://github.com/kean/ImagePublisher#going-from-low-to-high-resolution) should at least partially cover your requirements. And you can customize it any way you want. If you can't use Combine yet, there is a RxSwift version, or you can write it yourself.

Downsizing images locally is a bit more interesting. I used to do that in one of my previous apps where the downsizing logic was just what you described. I had some ideas about generalizing it in Nuke, but I haven't arrived at the final design yet.

In the current design, you can use `DataCacheOptions` to allow both original and processed image data to be stored in caches – this will cover the downsizing part. But there'll be no way to associate the downsized image with your server URL – it's designed for a different use case.

To fully cover your requirements, I would suggest creating a component that uses `ImagePipeline`, and that knows about your image sizes and automatically resizes the images in the completion callback. You can use `ImageProcessors.Resize` for resizing. To learn how write to caches, see https://github.com/kean/Nuke/issues/385.

**Comment by BB9z** [comment]
You can query the caching system to get a small size image, and then set it as the placeholder. I have some code like this:

```swift
private func cachedPreview(...: SomeModel) -> NSImage? {
    let request = ImageRequest(url: ...)
    if let image = pipeline.configuration.imageCache?[request]?.image {
        return image
    }
    let dcKey = pipeline.cacheKey(for: request, item: ...)
    if let data = pipeline.configuration.dataCache?.cachedData(for: dcKey) else {
        return NSImage(data: data)
    }
    return nil
}
```

**Comment by IsaiahJTurner** [comment]
@BB9z thanks that looks like exactly what we need. I will try!

I'll try to build this in a reusable way and I'll post the code back here when I'm done in case anyone else could benefit from it.

**Comment by kean** [comment]
I'm thinking about adding a high-level concept of [_responsive images_](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images), similar to what CSS offers:

```swift
<img srcset="elva-fairy-480w.jpg 480w,
             elva-fairy-800w.jpg 800w"
     sizes="(max-width: 600px) 480px,
            800px"
     src="elva-fairy-800w.jpg"
     alt="Elva dressed as a fairy">
```

The urgency to add it is not as high as it was for CSS because you as a user already has full control over the code. But it would be nice to add a convenience component that would take a simple set of rules (declarative) and apply them for you.

**Comment by IsaiahJTurner** [comment]
I'm not sure Nuke _needs_ responsive images either but there are a few features that could make responsive images infinitely better by being within Nuke itself:

1. The ability to set a duplicate request policy (i.e. wait for small request to finish before starting large request).
2. The ability to scale an image down to multiple cached sizes.
3. Nuke would run the fetches and render processing on the appropriate threads.

One oddity that I have to work around is that by using `pipeline.configuration.dataCache?.cachedData(for: diskCacheKey)` I need to manage my own DispatchWorkItems and ensure things run on the correct thread. I also have to handle timing issues and had to build on the solution above such that it checks if the image that will be requested via Nuke exists in the cache before calling `cachedData(for:)` and building a UIImage as that could duplicate the work Nuke does. I end up with a solution that barely uses Nuke at all for cache retrievals and duplicates work (e.g. checking FileExists).

**Comment by BB9z** [comment]
I have the same functional requirements. Regarding the design of the api interface, I have thought about it before, and the following is my draft.

We need to allow users to define two types of keys:

* Content key. Images with the same content key are considered the same and can be converted between different versions.
* Version key. Allow users to define different versions of a picture. This is necessary from a performance perspective. We should not allow high-resolution images to be scaled into small previews. For example, we may have a lot of small pictures in a list, but scaling large images will consumes too much memory and CPU time.

PS: I hope to use content key instead of the current `cacheKey`, `loadKey` or `filteredURL`.

**Comment by kean** [comment]
Thanks for the additional info! I think it's worth making improvements to the framework to (1) make it possible to implement these scenarios with no performance penalties and (2) provide a high-level API for some common scenarios. I'm not entirely sure about the high-level API because it will be challenging to not over-complicate it. But adding some lower-level APIs to cover (1) should be feasible.

**Comment by kean** [comment]
Hey, I spent some time designing the APIs to cover these scenarios and I would love to get some feedback on it. I'm pretty sure I already have all of these items in the [backlog](https://trello.com/b/Us4rHryT/nuke), but they are now taking more concrete form.

## 1. Add `ImageRequest.CachePolicy.returnCacheDataDontLoad`

`ImageRequest.CachePolicy` is an existing enum that currently has only two options: `default` and `reloadIgnoringCachedData`. I suggest adding `returnCacheDataDontLoad`.

This new policy will work similar to `URLRequest` [cachePolicy](https://developer.apple.com/documentation/foundation/urlrequest/cachepolicy) with the same name but will work with the custom data cache that you provide using `ImagePipeline.Configuration`. It will be cover the scenario where you want to quickly show a thumbnail (but only if it's available in the cache) while loading the original image.

## 2. Add Combine publishers

Make the code from [ImagePublisher](https://github.com/kean/ImagePublisher) part of the main framework to make it easier to install. There should now be enough demand to justify it.

Combine covers a lot of common scenarios, some of which I cover in [the documentation](https://kean.blog/nuke/guides/combine). For example, it covers "wait for small request to finish before starting large request.", and many more scenarios I can't predict.

## 3. Add `ImagePipelineDelegate`

I would argue that the scenario with multiple versions of the same image and multiple ways to produce these versions is uncommon. I'm not sure it justifies the addition of the new APIs specifically designed for it. Having said that, Nuke needs to provide a clear path to achieve this behavior.

What I suggest to do is add `ImagePipelineDelegate` to allow users to customize different stages of the pipeline based on the app's specific needs. In the initial version, it can allow users to customize the cache keys. Here is a draft:

```swift
/// Both requests produce the same image, but in two different ways.
func test()
        let requestA = ImageRequest(
            url: URL(string: "https://example.com/image-medium")!,
            processors: [ImageProcessors.Resize(width: 320)],
            cachePolicy: .returnCacheDataDontLoad,
            options: ImageRequestOptions(userInfo: ["contentKey": "Post-\(model.id)", "contentVersion": "small"])
        )

        let requestB = ImageRequest(
            url: URL(string: "https://example.com/image-small")!,
            processors: [], // No resizing needed
            options: ImageRequestOptions(userInfo: ["contentKey": "Post-\(model.id)", "contentVersion": "small"])
        )

        // TODO: use Combine to manage the requests. It can be serial, parallel, or whatever
        // kind of the execution the app needs.
    }
}

protocol ImagePipelineDelegate: AnyObject {
    func dataKey(for request: ImageRequest, kind: ImagePipeline.DataCacheItem) -> String
}

final class CustomImagePipelineDelegate: ImagePipelineDelegate {
    func dataKey(for request: ImageRequest, kind: ImagePipeline.DataCacheItem) -> String {
        switch kind {
        case .finalImage:
            guard let contentKey = request.options.userInfo["contentKey"] as? String,
                  let contentVersion = request.options.userInfo["contentVersion"] as? String else {
                return "" // TODO: provide a way to fallback to the default behavior
            }
            return contentKey + contentVersion
        case .originalImageData:
            return request.urlRequest.url?.absoluteString ?? ""
    }
}
```

## 4. Deprecate `ImageRequestOptions.cacheKey`, `loadKey`, and `filterURL`.

With `ImagePipelineDelegate`, there is no longer going to be a need for these options. I don't want to have two ways to achieve the same thing. These options were always a tad confusing and I was planning to get rid of them for a while – also one of the backlog items.

**Comment by BB9z** [comment]
> 1. Add ImageRequest.CachePolicy.returnCacheDataDontLoad
>
> This new policy will work similar to URLRequest cachePolicy with the same name but will work with the custom data cache that you provide using ImagePipeline.Configuration.

I'm a little confused. Isn't this function mentioned in Forward caching #385?

Let me re-describe a typical scenario. First, the user is accessing a list, and there are thumbnails of the content in the list. When the user clicks to enter the details, the big picture will start to load. When loading a large image, we want to display the loaded thumbnails in the list first to get a better user experience.

> 2. Add Combine publishers
>
> For example, it covers "wait for small request to finish before starting large request.", and many more scenarios I can't predict.

I have read that document before.  In my understanding this can combine two requests in series and wait for the final completion. But what we want is not simply to load the small image first, and then load the large image. 

In the scene I described above. When the big picture in the details starts to load, the loading of the big picture should be the first priority, and even the cancellation of the small picture is acceptable.

> 3. Add ImagePipelineDelegate

I prefer to set the cache key when making a request, rather than providing it in a delegate method.

1. We can do everything in one context, why should we divide it into two?
2. When creating a request, I can get more information. But in the delegate method, it is sometimes not easy to associate the ImageRequest object with the necessary information.

> 4. Deprecate ImageRequestOptions

Although it is not related to the current topic, I suggest that you can reduce the objects that need to be set when creating a request. Look at what we have: `ImageProcessing`, `ImageLoadingOptions`, `ImageRequestOptions`.

I have a wrapper for your reference

```swift
private struct RequestContext {
    var placeholder: NSImage?
    var fadeIn: TimeInterval = 0
    var pipeline: ImagePipeline?
    var processors = [ImageProcessing]()
    var cacheKey: String?
    var previewRequestCacheObject: PostModel?
    var imageView: NSImageView?
    var priority: ImageRequest.Priority = .normal
    var progress: ((_ intermediateResponse: ImageResponse?, _ completedUnitCount: Int64, _ totalUnitCount: Int64) -> Void)?
    var resultHandler: ((Result<ImageResponse, ImagePipeline.Error>) -> Void)?

    static func load(url: URL, _ config: (inout RequestContext) -> Void) -> ImageTask? {
        var context = RequestContext()
        config(&context)
        var option = ImageLoadingOptions.shared
        option.placeholder = context.placeholder
        if context.fadeIn > 0 {
            option.transition = .fadeIn(duration: context.fadeIn)
        }
        if let pipeline = context.pipeline {
            option.pipeline = pipeline
        }
        let processors = context.processors
        var request: ImageRequest!
        if let cacheRequest = context.previewRequestCacheObject?.cachedPreviewRequest as? ImageRequest {
            request = cacheRequest
        } else if let key = context.cacheKey {
            request = ImageRequest(url: url, processors: processors, options: ImageRequestOptions(cacheKey: key))
            if let requestHolder = context.previewRequestCacheObject {
                requestHolder.cachedPreviewRequest = request
            }
        } else {
            precondition(false)
        }
        request.priority = context.priority
        if let view = context.imageView {
            return Nuke.loadImage(with: request, options: option, into: view, progress: context.progress, completion: context.resultHandler)
        }
        return context.pipeline?.loadImage(with: request, progress: context.progress, completion: context.resultHandler)
    }
}
```

Use like

```swift
RequestContext.load(url: url) { cxt in
            cxt.pipeline = pipeline
            cxt.cacheKey = cacheKey(preview: post)
            cxt.imageView = view
            if inList {
                cxt.placeholder = NSImage.placeholder(post.site)
                cxt.fadeIn = 0.7
                cxt.processors.append(ImageProcessors.RoundedCorners(radius: 8))
            } else {
                cxt.priority = .high
            }
        }
```

**Comment by kean** [comment]
`Add ImageRequest.CachePolicy.returnCacheDataDontLoad` will cover the scenario where you want to check caches without performing networking. For example, display a thumbnail from cache while loading the full-size image. That should cover the @IsaiahJTurner use case.

> When the big picture in the details starts to load, the loading of the big picture should be the first priority, and even the cancellation of the small picture is acceptable.

You should be able to do that with Combine (or without it). There are too many possible scenarios to build them into the framework. For example, @IsaiahJTurner describes [this scenario](https://kean.blog/nuke/guides/combine#low-resolution-to-high): "wait for small request to finish before starting large request". Combine is perfectly well suited to cover all of them.

> I prefer to set the cache key when making a request, rather than providing it in a delegate method.

The design is still work-in-progress, but I'm leaning towards a delegate. Adding `contentKey` and `contentVersion` to the request will cover the specific use-case, but will complicate the public API for others. The delegate will give users more control and will arguably be easier to understand too. I personally prefer when a framework has fewer tools but offers great control so I can easily configure it to work exactly the way I want.

> Although it is not related to the current topic, I suggest that you can reduce the objects that need to be set when creating a request. Look at what we have: ImageProcessing, ImageLoadingOptions, ImageRequestOptions.

It looks similar to [ImageTaskBuilder](https://github.com/kean/ImageTaskBuilder), I would suggest checking it out. 

`ImageLoadingOptions` have a good reason to exist: it contains options that are only relevant to the UIKit extensions. There is no reason for the pipeline to know about them.

`ImageRequestOptions`, on the other hand, is just a drawer to hide some less useful options. I'm planning to re-evaluate it.

**Comment by kean** [comment]
[Nuke 9.6.0](https://github.com/kean/Nuke/releases/tag/9.6.0) shipped with `ImageRequest.CachePolicy.returnCacheDataDontLoad` and Combine publishes that are now built into the main repo. I'm working on further improvements to support the following scenarios in Nuke 10:

> Further, if the large size is available in the cache and we are attempting to display a small size, we'd like to cache that size separately for performance reasons by simply downsizing the large image without making a request to the server. Think, the native Photos app. Apple caches and stores thumbnails for all your photos in separate files so they don't need to load the full size image when scrolling thousands of photos very quickly.

**Comment by IsaiahJTurner** [comment]
`ImageRequest.CachePolicy.returnCacheDataDontLoad` this solves a lot of problems already.

I agree with @BB9z that delegates are not ideal although I don't think I would need to use this delegate anyway.

I would use the API like this:

```swift
// Cancellation is scoped to ImageRequest so we need to cancel the cache lookups if the request should be cancelled.
var isCancelled: Bool
// Size requests are sorted from ideal, to worst. 
let sizeRequests: [ImageRequest] = [
 ideal, // This will go over HTTP if all sizes are not found in cache.
 large,
 medium,
 small,
]

requestSize(sizeRequests: sizeRequests, index: 0) { image in
  print("Here's the best image we have so far!", image)
}

// Recursive function
func requestSize(sizeRequests: [ImageRequest], index: Int, progress: (UIImage?) -> Void) {
  let isHTTPRequest = index > sizeRequests.count
  var request = isHTTPRequest ? sizeRequests.first : sizeRequests[index]
  if isHTTPRequest {
    request.cachePolicy = .reloadIgnoringCachedData // a .usingMemoryCacheOnly would be better. Don't want to hit filesystem again here but technically a request could have already completed.
  } else {
    request.cachePolicy = .returnCacheDataDontLoad
  }
    pipeline?.loadImage(with: request, ...) { response in
      // progress (for progressive loading) and completion
      if NOT_FOUND_ERROR {
        requestSize(sizeRequests: sizeRequests, index: index + 1, progress: progress)
        return
      }
      progress(response.image)
    }
}
```

The only thing that this doesn't handle is avoiding duplicate requests in-flight and backfilling the memory caches if a different sized request finishes before the ideal request. I guess that's what combine is for but I haven't worked with combine yet. We've kinda worked around this sizing issue already by using `priority` to prioritize images displayed on the screen. We would never display two of the same size on the same page.

**Comment by IsaiahJTurner** [comment]
For reference, this is what we ended up with and are using in production at the moment. It is just a copy paste so not super easy to understand but we basically load groups of 1-5 images at once (a gif, but separately because gifs aren't progressive).

```swift
		private func cachedImageForImageRequest(_ imageRequest: ImageRequest) -> UIImage? {
			let pipeline = self.mediaCache.pipeline
			if let image = pipeline.configuration.imageCache?[imageRequest]?.image {
				return image
			}

			let diskCacheKey = pipeline.cacheKey(for: imageRequest, item: .finalImage)
			if let data = pipeline.configuration.dataCache?.cachedData(for: diskCacheKey) {
				return UIImage(data: data)
			}
			return nil
		}
		init(mediaCache: MediaCache, urls: [URL], size: Size, quality: Quality?, progressHandler: @escaping (Task) -> Void) {
			self.progressHandler = progressHandler
			self.mediaCache = mediaCache
			self.urls = urls
			self.internalImages = Array(repeating: nil, count: urls.count)
			self.tasks = [ImageTask]()
			self.size = size
			self.quality = quality
			for (index, url) in urls.enumerated() {
				let scopedIndex = index
				guard let imageRequest = mediaCache.imageRequestForURL(url, size: size, quality: quality) else {
					self.cancel()
					self.isFailed = true
					return
				}
				self.imageRequests.append(imageRequest)
				let task = mediaCache.pipeline.loadImage(with: imageRequest, queue: imageQueue, progress: { [weak self] (response, completed, total) in
					guard let response = response else {
						return
					}
					self?.internalImages[scopedIndex] = response.image
				}, completion: { [weak self] (result) in
					switch result {
					case .failure(let error):
						Log.error(error)
						self?.isFailed = true
					case .success(let response):
						self?.tasksCompleted += 1
						self?.internalImages[scopedIndex] = response.image
					}
				})
				self.tasks.append(task)
				
				imageCacheQueue.async {
					if let filename = self.mediaCache.dataCache?.filename(for: self.mediaCache.pipeline.cacheKey(for: imageRequest, item: .originalImageData)) {
						
						if let fileURL = self.mediaCache.dataCache?.path.appendingPathComponent(filename),
						   FileManager.default.fileExists(atPath: fileURL.path) {
							return
						}
					}
					let cachedSizes: [Size] = [
						.large,
						.medium,
						.original,
						.small,
					].filter { $0 != size }
					for unscoped in urls.enumerated() {
						let url = unscoped.element
						let index = unscoped.offset
						var cachedImage: UIImage?
						cachedSizes.forEach { cachedSize in
							if cachedImage != nil {
								return
							}
							if let imageRequest = mediaCache.imageRequestForURL(url, size: cachedSize, quality: quality) {
								cachedImage = self.cachedImageForImageRequest(imageRequest)
							}
						}
						imageQueue.async {
							if self.internalImages[index] == nil {
								self.internalImages[index] = cachedImage
							}
						}
					}
				}
			}
		}
```

**Comment by kean** [comment]
> The only thing that this doesn't handle is avoiding duplicate requests in-flight and backfilling the memory caches if a different sized request finishes before the ideal request.

Nuke has pretty smart [coalescing](https://kean.blog/nuke/guides/image-pipeline-guide#coalescing) for requests built-in. That's what `ImagePrefetcher` also relies on it – Nuke is designed to be used this way. So in most cases you won't see any duplicate requests.

The code looks good. You don't really need Combine to achieve what you need.

> backfilling the memory caches

You can use `ImagePrefetcher` or simply send these requests yourself with a lower [priority](https://kean.blog/nuke/guides/customizing-requests#priority).

**Comment by kean** [comment]
`.usingMemoryCacheOnly` – I'm adding a set of more granular controls in `ImageRequest` in Nuke 10, along with many other cache-related changes.

**Comment by IsaiahJTurner** [comment]
@kean does `returnCacheDataDontLoad` return progressive images that haven't finished loading yet?

**Comment by kean** [comment]
Progressive scans are stored only in memory cache and only if `isStoringPreviewsInMemoryCache` is enabled (will be enabled by default in Nuke 10). Yes, you'll receive the scan in the `progress` handler if it's available in the memory cache.

**Comment by kean** [comment]
The changes I outlined earlier are now available in [Nuke 10 Beta](https://github.com/kean/Nuke/releases/tag/10.0.0-beta.1). I would appreciate your feedback.

I'm going to close this issue because it's already mostly addressed and it's getting a bit hard to follow. If you need more APIs, please feel free to open a new one.
